jocktos/inc/synchro.h

```diff
@@ -0,0 +1,74 @@
+/**
+* \brief This header is to act as companion header for synchro.c
+*/
+#ifndef _SEMAPHORE_H_
+#define _SEMAPHORE_H_
+/* -- Includes ------------------------------------------------------------ */
+// Jocktos
+#include "tcb.h"
+// Middleware
+// Bios
+// Standard C
+#include <stdint.h>
+
+/* -- Defines ------------------------------------------------------------- */
+
+/**
+ * @brief default semaphore is binary (mutex)
+ */
+#define T_SEMAPHORE_DEF(...) \
+{   /* ---Internal Data---*/  \
+    .value_           = 1,    \
+    .count            = 2,    \
+    .pendingTCBQueue_ = NULL, \
+    .ownersPriority_  = 0,    \
+     __VA_ARGS__              \
+}
+
+/* -- Types --------------------------------------------------------------- */
+
+/**
+ *  @brief Represents a semaphore for controlling access to a shared resource. 
+ */
+typedef struct {
+    uint16_t value_;                                ///< [INTERNAL] current value

```

Comment:
What if we were to hide some of these private members?

Say if we gave them an opaque handle (void * with a pretty name) and provide inits/destroyer and accessor functions for everything...
```suggestion
struct SomeStruct {
int member;
void* internals; //allocate this to your private struct
};
```
Like so. 
 
If people want to screw up, they should be allowed to. It is C after all

Reply:
That's interesting, I was actually considering something similar for the task functions where tasks could access a subset of tcb properties for advanced features like self canceling, priority changes etc... 

Ultimately I figured the cast back from void* and 2nd layer of references was too much overhead for security / encapsulation since this hasn't been a focus for JOCKTOS so far. 

Definitely open to reconsider this though, it could add some real value to the project overall